File: backend/ai_parser.js
const Groq = require("groq-sdk");

const groq = new Groq({
    apiKey: process.env.GROQ_API_KEY
});

const GROQ_MODEL = "openai/gpt-oss-20b";


// ======================================================
// CLEAN AI JSON RESPONSE
// ======================================================
function cleanAIResponse(responseText) {
    if (!responseText) {
        throw new Error("AI returned an empty response.");
    }

    let cleaned = responseText.trim();

    cleaned = cleaned
        .replace(/^```json\s*/i, "")
        .replace(/^```\s*/i, "")
        .replace(/\s*```$/i, "")
        .trim();

    return cleaned;
}


// ======================================================
// PARSE JSON SAFELY
// ======================================================
function parseAIJson(responseText) {
    const cleaned = cleanAIResponse(responseText);

    try {
        return JSON.parse(cleaned);
    } catch (error) {
        const firstBrace = cleaned.indexOf("{");
        const lastBrace = cleaned.lastIndexOf("}");

        if (firstBrace !== -1 && lastBrace > firstBrace) {
            const possibleJson = cleaned.slice(
                firstBrace,
                lastBrace + 1
            );

            try {
                return JSON.parse(possibleJson);
            } catch (_) {
                // Fall through to the useful error below.
            }
        }

        throw new Error(
            "AI returned invalid JSON."
        );
    }
}


// ======================================================
// AI RESUME PARSER
// ======================================================
async function parseResumeWithAI(resumeText) {
    if (!resumeText || !resumeText.trim()) {
        throw new Error(
            "No resume text was provided."
        );
    }

    if (!process.env.GROQ_API_KEY) {
        throw new Error(
            "GROQ_API_KEY is missing from the backend .env file."
        );
    }

    console.log("🤖 Sending resume to Groq...");

    const prompt = `
You are an expert resume parser.

Analyze the resume text below and extract information into the EXACT JSON structure provided.

IMPORTANT RULES:
1. Extract information based ONLY on the resume.
2. NEVER invent information.
3. If information is missing, use an empty string "".
4. If a section has multiple entries, return ALL entries.
5. Understand different resume formats and section names.
6. Do not depend on sections appearing in a specific order.
7. Correctly identify dates, durations and locations.
8. Keep descriptions meaningful but concise.
9. Extract URLs from the resume.
10. Separate skills individually.
11. Distinguish projects from work experience.
12. Distinguish achievements from certifications.
13. Do not confuse a job title with the person's name.
14. Return ONLY valid JSON.
15. Do not use markdown.
16. Do not add explanations.
17. Do not invent a professional introduction. If the resume has no summary/profile/about section, return "".
18. Keep all facts faithful to the resume.

JSON STRUCTURE:
{
    "name": "",
    "email": "",
    "phone": "",
    "location": "",
    "introduction": "",
    "education": [
        {
            "degree": "",
            "institution": "",
            "duration": ""
        }
    ],
    "experience": [
        {
            "role": "",
            "company": "",
            "duration": "",
            "description": ""
        }
    ],
    "projects": [
        {
            "name": "",
            "description": "",
            "technologies": ""
        }
    ],
    "skills": [],
    "achievements": [],
    "certifications": [],
    "github": "",
    "linkedin": ""
}

RESUME TEXT:
${resumeText}
`;

    try {
        const completion = await groq.chat.completions.create({
            messages: [
                {
                    role: "system",
                    content:
                        "You are a highly accurate resume information extraction system. Return only valid JSON."
                },
                {
                    role: "user",
                    content: prompt
                }
            ],
            model: GROQ_MODEL,
            temperature: 0,
            max_tokens: 4000
        });

        const responseText =
            completion?.choices?.[0]?.message?.content?.trim();

        console.log("🤖 Groq raw response:");
        console.log(responseText);

        const parsedData = parseAIJson(responseText);

        console.log("✅ Resume successfully parsed by AI!");

        return parsedData;
    } catch (error) {
        console.error(
            "❌ AI resume parsing failed:",
            error.message || error
        );
        throw error;
    }
}


// ======================================================
// AI CAREER DATA IMPROVEMENT
// ======================================================
async function improveCareerDataWithAI(careerData) {
    if (!careerData) {
        throw new Error(
            "No career data was provided."
        );
    }

    if (!process.env.GROQ_API_KEY) {
        throw new Error(
            "GROQ_API_KEY is missing from the backend .env file."
        );
    }

    console.log(
        "✨ Sending career data to Groq for improvement..."
    );

    const prompt = `
You are an expert professional resume and portfolio writer.

Improve ONLY the user's professional introduction and project descriptions.

IMPORTANT RULES:
1. Use ONLY information already present in the provided data.
2. NEVER invent skills, companies, achievements, technologies, jobs, education, responsibilities or experience.
3. Do not change factual information.
4. Make the writing professional, clear and concise.
5. Make the introduction stronger while keeping it truthful.
6. Improve project descriptions so they clearly explain what was built and what problem it addresses, but only using facts already present.
7. Keep the original meaning.
8. If introduction is empty, return an empty string.
9. If projects are empty, return an empty array.
10. Do not modify project names.
11. Do not modify technologies.
12. Do not modify education.
13. Do not modify experience facts.
14. Do not invent metrics or results.
15. Return ONLY valid JSON.
16. No markdown.
17. No explanation.

Return EXACTLY this structure:
{
    "improvedIntroduction": "",
    "improvedProjects": [
        {
            "name": "",
            "originalDescription": "",
            "improvedDescription": ""
        }
    ]
}

CAREER DATA:
${JSON.stringify(careerData, null, 2)}
`;

    try {
        const completion = await groq.chat.completions.create({
            messages: [
                {
                    role: "system",
                    content:
                        "You are a professional portfolio improvement assistant. Return only valid JSON."
                },
                {
                    role: "user",
                    content: prompt
                }
            ],
            model: GROQ_MODEL,
            temperature: 0.2,
            max_tokens: 2500
        });

        const responseText =
            completion?.choices?.[0]?.message?.content?.trim();

        console.log("🤖 AI improvement response:");
        console.log(responseText);

        const improvedData = parseAIJson(responseText);

        console.log("✅ AI improvements generated.");

        return improvedData;
    } catch (error) {
        console.error(
            "❌ AI improvement failed:",
            error.message || error
        );
        throw error;
    }
}


// ======================================================
// EXPORT
// ======================================================
module.exports = {
    parseResumeWithAI,
    improveCareerDataWithAI
};

File: backend/ai_parser.py
import os
import json

from dotenv import load_dotenv
from groq import Groq

from resume_parser import extract_text_from_pdf, clean_text


# Find the .env file inside the backend folder
env_path = os.path.join(os.path.dirname(__file__), ".env")

load_dotenv(env_path)

api_key = os.getenv("GROQ_API_KEY")


if not api_key:
    print("Groq API key NOT found!")
    exit()

client = Groq(api_key=api_key)


def parse_resume(pdf_path):

    # Extract text from PDF
    resume_text = extract_text_from_pdf(pdf_path)

    # Clean extracted text
    resume_text = clean_text(resume_text)

    # Send resume text to Groq
    response = client.chat.completions.create(
        model="openai/gpt-oss-20b",
        temperature=0,

        messages=[
            {
                "role": "system",
                "content": """
You are a professional resume parser.

Extract ONLY information that is explicitly present in the resume.

Rules:
1. Never invent or assume information.
2. Return ONLY valid JSON.
3. Every required field must be included.
4. Missing single-value fields must be null.
5. Missing list fields must be [].
6. Do not add extra fields.
7. Preserve information exactly where possible.
"""
            },
            {
                "role": "user",
                "content": f"Parse this resume:\n\n{resume_text}"
            }
        ],

        response_format={
            "type": "json_schema",
            "json_schema": {
                "name": "resume",
                "strict": True,

                "schema": {
                    "type": "object",

                    "properties": {

                        "name": {
                            "type": ["string", "null"]
                        },

                        "email": {
                            "type": ["string", "null"]
                        },

                        "phone": {
                            "type": ["string", "null"]
                        },

                        "location": {
                            "type": ["string", "null"]
                        },

                        "summary": {
                            "type": ["string", "null"]
                        },

                        "skills": {
                            "type": "array",
                            "items": {
                                "type": "string"
                            }
                        },

                        "education": {
                            "type": "array",
                            "items": {
                                "type": "object",

                                "properties": {

                                    "degree": {
                                        "type": ["string", "null"]
                                    },

                                    "institution": {
                                        "type": ["string", "null"]
                                    },

                                    "location": {
                                        "type": ["string", "null"]
                                    },

                                    "start_year": {
                                        "type": ["string", "null"]
                                    },

                                    "end_year": {
                                        "type": ["string", "null"]
                                    }
                                },

                                "required": [
                                    "degree",
                                    "institution",
                                    "location",
                                    "start_year",
                                    "end_year"
                                ],

                                "additionalProperties": False
                            }
                        },

                        "projects": {
                            "type": "array",
                            "items": {
                                "type": "object",

                                "properties": {

                                    "title": {
                                        "type": "string"
                                    },

                                    "description": {
                                        "type": "string"
                                    },

                                    "technologies": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                },

                                "required": [
                                    "title",
                                    "description",
                                    "technologies"
                                ],

                                "additionalProperties": False
                            }
                        },

                        "experience": {
                            "type": "array",
                            "items": {
                                "type": "object",

                                "properties": {

                                    "job_title": {
                                        "type": "string"
                                    },

                                    "company": {
                                        "type": "string"
                                    },

                                    "location": {
                                        "type": ["string", "null"]
                                    },

                                    "start_date": {
                                        "type": ["string", "null"]
                                    },

                                    "end_date": {
                                        "type": ["string", "null"]
                                    },

                                    "description": {
                                        "type": "string"
                                    }
                                },

                                "required": [
                                    "job_title",
                                    "company",
                                    "location",
                                    "start_date",
                                    "end_date",
                                    "description"
                                ],

                                "additionalProperties": False
                            }
                        },

                        "certifications": {
                            "type": "array",
                            "items": {
                                "type": "string"
                            }
                        },

                        "achievements": {
                            "type": "array",
                            "items": {
                                "type": "string"
                            }
                        },

                        "links": {
                            "type": "array",
                            "items": {
                                "type": "object",

                                "properties": {

                                    "name": {
                                        "type": "string"
                                    },

                                    "url": {
                                        "type": "string"
                                    }
                                },

                                "required": [
                                    "name",
                                    "url"
                                ],

                                "additionalProperties": False
                            }
                        }
                    },

                    "required": [
                        "name",
                        "email",
                        "phone",
                        "location",
                        "summary",
                        "skills",
                        "education",
                        "projects",
                        "experience",
                        "certifications",
                        "achievements",
                        "links"
                    ],

                    "additionalProperties": False
                }
            }
        }
    )

    # Convert JSON string into Python dictionary
    resume_data = json.loads(
        response.choices[0].message.content
    )

    return resume_data


# Test the parser
if __name__ == "__main__":

    print("Groq API key loaded successfully!")

    pdf_path = os.path.join(
        os.path.dirname(__file__),
        "..",
        "sample_resumes",
        "resume2.pdf"
    )

    resume_data = parse_resume(pdf_path)

    print("\n===== PARSED RESUME =====")

    print("Name:", resume_data["name"])
    print("Email:", resume_data["email"])
    print("Phone:", resume_data["phone"])
    print("Location:", resume_data["location"])
    print("Summary:", resume_data["summary"])

    print("Skills:", resume_data["skills"])
    print("Education:", resume_data["education"])
    print("Projects:", resume_data["projects"])
    print("Experience:", resume_data["experience"])
    print("Certifications:", resume_data["certifications"])
    print("Achievements:", resume_data["achievements"])
    print("Links:", resume_data["links"])